Astropy model fitting

KriSun95 opened the pull request:
This PR aims to start introducing code in the fitting refactor, first aiming to implement `astropy` models (discussion started in #81).

A few scripts are included into the `fitting` module and split into sub-modules. These sub-modules revolve around the models being used to fit data, general objective functions, somewhere for different statistics, and one for optimiser tools.

The thought process behind the modules is:
- `models`: Somewhere to keep all the cool models we want. 
- `statistics`: The functions in here should only take in the data and the model arrays that need to be compared (errors eventually, etc). This module could have a bunch of different statistics (e.g., Gaussian, Poissonian, etc.).
- `objective_functions`: these hold the functions that are actually being minimised. they take in the information to evaluate a given model at a bunch of different parameters and pass the model result over to whatever fit statistic you want.
- `optimizer_tools`(should be `optimiser_tools` really!): Handles the interface between the objective function and whatever fitting routine is used. In the case here, `script.minimize`.

An `example` folder is created that includes a very simple example script showing the generation of fake data to be fitted, a dummy instrument response matrix, and the fitting of that fake data using `scipy` and `astropy`. The results are also plotted in the script and the result should be the following
![astropy-fitting-example](https://github.com/sunpy/sunkit-spex/assets/43237137/63b10185-d651-43a4-85dc-60a9c9713b07)

Feel free to play around with the code and any feedback is encouraged!

Next steps could include, but are not limited to
- dealing with non-square instrument responses 
- simultaneous fitting

**Note:** An additional change was implemented in `sunkit_spex/legacy/fitting_legacy/fitter.py` as `numpy` has moved their warnings to another model in v2, I think.

samaloney commented on sunkit_spex/fitting/examples/simple_fitting.py:
```diff
@@ -0,0 +1,304 @@
+from sunkit_spex.fitting.statistics.gaussian import chi_squared
+
+# decide whether to save the figure at the end
+SAVE_FIG = False
```
Not needed in sphinx-gallery

samaloney commented on sunkit_spex/fitting/examples/simple_fitting.py:
```diff
@@ -0,0 +1,304 @@
+                             title="Fake Count Spectrum"):
+    """ Plots the fake count spectrum and its componenets. """
+
+    axis.plot(photon_energies,
+              photon_model_features,
+              label="photon model features")
+    axis.plot(photon_energies,
+              gaussian_feature,
+              label="gaussian feature")
+    axis.plot(photon_energies,
+              total_count_spectrum,
+              label="total fake spectrum")
+    axis.plot(photon_energies,
+              total_count_spectrum_wnoise,
+              label="total fake spectrum + noise",
+              lw=0.5)
```
I think you should be able to iterate over the model components

KriSun95 replied:
I thought so too when I wrote this but was too lazy :/

I've re-ordered the example for the docs now (which I think looks and flows way better now) so I don't think this is as much of an issue with the written code now. I could be wrong though!

samaloney commented on sunkit_spex/fitting/models/__init__.py:
All the models should be in `sunkit_spex.models`

samaloney reviewed (changes requested):
A great start! I think a few things need to be added/updated:
1. Should add tests the example gallery will run the code but it doesn't replace tests
2. The code needs to be added to the documentation

KriSun95 commented on sunkit_spex/fitting/models/data/gaussian.py:
```diff
@@ -0,0 +1,15 @@
+from astropy.modeling import Fittable1DModel, Parameter
+
+
+class GaussianCountModel(Fittable1DModel):
```
I don't think it's necessary here but I was unsure when it came to including units and/or different energy grids. I'll combine them for now and we can separate them if/when we have to.

KriSun95 commented on sunkit_spex/fitting/examples/simple_fitting.py:
```diff
@@ -0,0 +1,304 @@
+# get some fake photon energies
+
+
+def photon_energies(start, stop, inc):
+    """ Get a `ndarray` of energies. """
+    return np.arange(start, stop, inc)
```
No. I thought it would be more useful but it turned out to be pretty much useless and I forgot to remove it.

KriSun95 commented on sunkit_spex/fitting/examples/simple_fitting.py:
```diff
@@ -0,0 +1,304 @@
+    axis.plot(photon_energies, photon_values)
+    axis.set_xlabel("Energy [keV]")
+    axis.set_ylabel("ph s$^{-1}$ cm$^{-2}$ keV$^{-1}$")
+    axis.set_title(title)
```
Yeah, I think that would be a really good goal when the spectrum object is introduced! Would this be something to introduce in the next PR or try to add more into this one?

KriSun95 commented:
Quite a few things have been updated. I focussed on addressing the suggestions made by @samaloney which has put the code in a much better spot, I think. I've also added in a number of tests for the stuff I wrote as well.

samaloney commented on docs/sg_execution_times.rst:
I don't think this should be in the repo as it will be generated on read the docs (RTD)

KriSun95 replied:
Deleted the file and added to the gitignore.

samaloney reviewed (approved):
Looks good to me except for the rouge `sg_execution_times.rst` file.

samaloney commented:
@hayesla @DanRyanIrish could one of you review so we can maybe get this merged

Cadair commented:
I don't know where the best place to discuss this is, but I have been working on improving the performance of astropy modelling fitters and implemented a [helper function for parallel fitting](https://docs.astropy.org/en/latest/modeling/parallel-fitting.html) I would be super interested to hear your thoughts on what would need to be done to make this useful for your use cases.

DanRyanIrish commented:
I think the `StraightLineModel` and `GaussianModel` seem like placeholder classes.  Unless they gain more functionality than the astropy versions, they should be removed in the future.  But I don't want to hold this up getting merged.
